server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/InternalHistogram.java

```diff
@@ -347,56 +348,75 @@ private double round(double key) {
     }
 
     private void addEmptyBuckets(List<Bucket> list, ReduceContext reduceContext) {
-        ListIterator<Bucket> iter = list.listIterator();
+        class Counter implements DoubleConsumer {
+            private int size = list.size();
+
+            @Override
+            public void accept(double key) {
+                size++;
+                if (size >= 10000) {
+                    reduceContext.consumeBucketsAndMaybeBreak(size);
+                    size = 0;
+                }
+            }
+        }
+        Counter counter = new Counter();
+        iterateEmptyBuckets(list, list.listIterator(), counter);
```

Comment:
If I followed this correctly, this pass doesn't actually allocate the buckets, it just runs the maybe break check based on how many buckets we would allocate, and the call on line 371 does the actual allocation.  Presuming I followed that correctly, I think it's worth leaving a comment to that effect.  That seems to be the core of this fix, and it'd be good to remember why we're doing it this way.

Reply:
I'm happy to do that.

@imotov I'd love to hear from you too on this PR. To get two green check marks on this one.